### ISSUE TITLE
管道demo

### ISSUE THREAD

**Issue by holdyounger** [holdyounger/Issue#53]
管道demo

[toc]

# 管道类

##  .h

~~~cpp
#pragma once

#include &lt;stdio.h&gt;
#include &lt;tchar.h&gt;

#include &lt;iostream&gt;
#include &lt;Windows.h&gt;
#include &lt;string&gt;
#include &lt;thread&gt;
#include &lt;memory&gt;

#define TIME_ASECOND 1000

enum PIPEUSERTYPE {
	USER_CLIENT, //客户端
	USER_SERVER, //服务端
};

class PipeIPC {
public:
	explicit PipeIPC(PIPEUSERTYPE role) :m_Role(role), m_hPipeHandle(INVALID_HANDLE_VALUE) {};
	~PipeIPC();

	bool WriteData(__in const std::string&amp; datas); //写入数据
	void Release();
	bool ReadData(__out std::string&amp; datas, __out bool bIsPipeEnd); //读取数据
	bool InitPipeIPC(); //初始化管道通信

	bool IsInit();

	BOOL InitSecurityAttributes();

	void Connect();
	void DisConnect();
private:
	PIPEUSERTYPE m_Role;
	HANDLE m_hPipeHandle;
	SECURITY_ATTRIBUTES m_sec_attr;  //security attributes 
};
~~~

## .cpp

~~~cpp
#include &quot;PipeIPC.h&quot;
#include &lt;AccCtrl.h&gt;
#include &lt;AclAPI.h&gt;

using namespace std;

#define PIPE_SERVER_NAME       &quot;&#92;&#92;&#92;&#92;.&#92;&#92;pipe&#92;&#92;TrustAgent&#92;&#92;pipe.ToFixService&quot;
#define PIPE_CLIENT_NAME       &quot;&#92;&#92;&#92;&#92;.&#92;&#92;pipe&#92;&#92;TrustAgent&#92;&#92;pipe.ToTrustAgent&quot;

PipeIPC::~PipeIPC()
{
	if (m_hPipeHandle != INVALID_HANDLE_VALUE)
	{
		DisconnectNamedPipe(m_hPipeHandle);
		CloseHandle(m_hPipeHandle);
	}
}

bool PipeIPC::InitPipeIPC()
{
	if (USER_SERVER == m_Role &amp;&amp; nullptr != m_hPipeHandle)
	{
		InitSecurityAttributes();
		// cout &lt;&lt; &quot;ser&quot; &lt;&lt; endl;
		m_hPipeHandle = CreateNamedPipeA(PIPE_SERVER_NAME, PIPE_ACCESS_DUPLEX, //双向模式
			PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE/* | PIPE_NOWAIT*/,
			PIPE_UNLIMITED_INSTANCES,//设置最大容纳实例数目
			0,
			0,
			NULL, (LPSECURITY_ATTRIBUTES)&amp;m_sec_attr);
		//如果管道创建失败
		if (m_hPipeHandle == INVALID_HANDLE_VALUE)
		{
			return false;
		}

#if 0
		if (!ConnectNamedPipe(m_hPipeHandle, nullptr)) {
			cout &lt;&lt; &quot;ConnectNamedPipe failed&quot; &lt;&lt; endl;
			return false;
		}
#endif
	}
	else
	{
		if (WaitNamedPipeA(PIPE_CLIENT_NAME, TIME_ASECOND * 5) == FALSE)
			return 0;

		m_hPipeHandle = CreateFileA(
			PIPE_CLIENT_NAME,
			GENERIC_READ | GENERIC_WRITE,
			0,
			NULL,
			OPEN_EXISTING,
			FILE_ATTRIBUTE_NORMAL,
			NULL);

		//如果管道创建失败
		if (m_hPipeHandle == INVALID_HANDLE_VALUE)
		{
			return false;
		}

		WriteData(&quot;[PIPE] ------ V10 Init ------&quot;);
	}

	return true;
}

bool PipeIPC::IsInit()
{
	return m_hPipeHandle == INVALID_HANDLE_VALUE ? false : true;
}

/*
	[details] 创建管道去管理员权限
*/
BOOL PipeIPC::InitSecurityAttributes()
{
	PSID pEveryoneSID = NULL;
	PACL pACL = NULL;
	EXPLICIT_ACCESS ea;
	PSECURITY_DESCRIPTOR pSD = NULL;
	void* ea_addr = NULL;
	void* acl_addr = NULL;
	void* sd_addr = NULL;
	SID_IDENTIFIER_AUTHORITY SIDAuthWorld = SECURITY_WORLD_SID_AUTHORITY;

	if (!AllocateAndInitializeSid(&amp;SIDAuthWorld, 1, SECURITY_WORLD_RID, 0, 0, 0, 0, 0, 0, 0, &amp;pEveryoneSID))
	{
		return FALSE;
	}
	ZeroMemory(&amp;ea, sizeof(EXPLICIT_ACCESS));
	ea.grfAccessPermissions = SPECIFIC_RIGHTS_ALL | STANDARD_RIGHTS_ALL;
	ea.grfAccessMode = SET_ACCESS;
	ea.grfInheritance = NO_INHERITANCE;
	ea.Trustee.TrusteeForm = TRUSTEE_IS_SID;
	ea.Trustee.TrusteeType = TRUSTEE_IS_WELL_KNOWN_GROUP;
	ea.Trustee.ptstrName = (LPTSTR)pEveryoneSID;
	ea_addr = &amp;ea;
	if (ERROR_SUCCESS != SetEntriesInAcl(1, &amp;ea, NULL, &amp;pACL))
	{
		goto cleanup;
	}
	acl_addr = &amp;pACL;
	pSD = (PSECURITY_DESCRIPTOR)LocalAlloc(LPTR, SECURITY_DESCRIPTOR_MIN_LENGTH);
	if (NULL == pSD)
	{
		goto cleanup;
	}
	if (!InitializeSecurityDescriptor(pSD, SECURITY_DESCRIPTOR_REVISION))
	{
		goto cleanup;
	}
	sd_addr = &amp;pSD;
	if (!SetSecurityDescriptorDacl(pSD, TRUE, pACL, FALSE))
	{
		goto cleanup;
	}
	m_sec_attr.nLength = sizeof(SECURITY_ATTRIBUTES);
	m_sec_attr.lpSecurityDescriptor = pSD;
	m_sec_attr.bInheritHandle = FALSE;
	return TRUE;

cleanup:
	if (pEveryoneSID)
		FreeSid(pEveryoneSID);
	if (pACL)
		LocalFree(pACL);
	if (pSD)
		LocalFree(pSD);
	return FALSE;
}

void PipeIPC::Connect()
{
	ConnectNamedPipe(m_hPipeHandle, nullptr);
}

void PipeIPC::DisConnect()
{
	DisconnectNamedPipe(m_hPipeHandle);
}

//读取数据
bool PipeIPC::ReadData(__out std::string&amp; datas, __out bool bIsPipeEnd)
{
	if (INVALID_HANDLE_VALUE == m_hPipeHandle || nullptr == m_hPipeHandle)
	{
		InitPipeIPC();
	}

	DWORD cbRead = 0;//定义数据长度
	DWORD cbData = 0;
	if (ReadFile(m_hPipeHandle, &amp;cbData, sizeof(cbData), &amp;cbRead, nullptr) &amp;&amp; sizeof(cbData) == cbRead)
	{
		DWORD error = GetLastError();
		bIsPipeEnd = (error == ERROR_BROKEN_PIPE ? true : false);

		datas.resize(cbData);

		if (ReadFile(m_hPipeHandle, (LPVOID)datas.data(), cbData, &amp;cbRead, nullptr) &amp;&amp; cbData == cbRead)
		{
			return true;
		}
	}
	else
	{
		DWORD dwError = GetLastError();
		bIsPipeEnd = (dwError == ERROR_BROKEN_PIPE) ? true : false;
	}

	datas.resize(0);
	return false;
}

//写入数据
bool PipeIPC::WriteData(__in const std::string&amp; datas)
{
	if (m_Role == USER_SERVER)
	{
		bool connect = ConnectNamedPipe(m_hPipeHandle, NULL);
		if (!connect)
		{
			if (!InitPipeIPC())
			{
				DWORD error = GetLastError();
				if (error != ERROR_PIPE_CONNECTED)
				{
					//cout &lt;&lt; &quot;客户端还没连接&quot; &lt;&lt; endl;
					//连接失败
					return false;
				}
			}
		}
	}

	if (!IsInit())
	{
		if (!InitPipeIPC())
		{
			return false;
		}
	}

	//连接成功！
	DWORD wLen;
	const DWORD cbData = datas.size() * sizeof(datas[0]);

	if (!WriteFile(m_hPipeHandle, &amp;cbData, sizeof(cbData), &amp;wLen, NULL))
	{
		return false;
	}
	if (!WriteFile(m_hPipeHandle, datas.c_str(), datas.size() * sizeof(datas[0]), &amp;wLen, NULL))
	{
		return false;
	}

	return true;
}

void PipeIPC::Release()
{
	if (m_hPipeHandle)
	{
		DisconnectNamedPipe(m_hPipeHandle);
		CloseHandle(m_hPipeHandle); // 关闭句柄
		m_hPipeHandle = NULL;
	}
}
~~~

# 定义与声明

~~~cpp
__declspec(selectany) PipeIPC* pRead = new PipeIPC(USER_SERVER);
__declspec(selectany) PipeIPC* pWrite = new PipeIPC(USER_CLIENT);
~~~

# 使用

## 监听管道

&gt;  随程序同时启动

~~~cpp
   
BOOL StartPipeRead()
{
	EventLog(LOG_ALL, &quot;[IPC][THREAD] ,Create PipeReadThread Success, pipe = %p&quot;, pRead );
	EventLog(LOG_ALL, &quot;pipe = %p, pipe-&gt;IsInit()=%d&quot;, pRead, pRead-&gt;IsInit());

	bool isPipeEnd = true;
	while (true)
	{
		if (isPipeEnd)
		{
			pRead-&gt;Connect();
		}

		std::string datas;
		if (!pRead-&gt;ReadData(datas, isPipeEnd))
		{
			if (isPipeEnd)
			{
				EventLog(LOG_ALL, &quot;-----------[PIPE END]----------&quot;);
				pRead-&gt;DisConnect();
			}
		}

		datas = &quot;Add&quot;;

		// 数据处理
		if (datas.find(&quot;start&quot;) != std::string::npos) // 启动成功
		{
			EventLog(LOG_ALL, &quot; [PIPE] [REC] TrustAgentStart&quot;);
			// 启动 管道客户端
			if (!StartPipeWrite())
			{
				EventLog(LOG_ALL, &quot; [PIPE] Failed to Create Client Pipe, GetLastError()=&quot;, GetLastError());
			}
		}
		else if (!datas.empty())
		{
			EventLog(LOG_ALL, &quot;[PIPE] [RECV] datas: %s&quot;, datas.c_str());
			// 调用导出的接口
			{
				std::thread th(HandleFix, datas);
				th.detach();

			}
			// NSPTRUST::SendRpcMsgToPlugin(POLICY_SENDTQA_USER, datas.c_str());
		}
	}
}
~~~

## 写管道

启动时机，同名管道的监听管道启动后再启动

~~~cpp
BOOL StartPipeWrite()
{
	if (pWrite-&gt;IsInit())
		return true;
	else
		return pWrite-&gt;InitPipeIPC();
}
~~~

写数据

~~~cpp
	std::string datas = &quot;Send data to PIPE Read&quot;;
	pWrite-&gt;WriteData(datas);
~~~

> blog link [管道demo](https://holdyounger.github.io/OS/Windows/管道的使用/)
